File: src/bin/liq_kamino_live.rs
//! Kamino LIVE-health finder — recomputes each obligation's health from CURRENT
//! reserve prices (replicating refresh_obligation), instead of trusting the
//! obligation's stored (possibly stale) values.
//!
//! Two outputs:
//!   1. VALIDATION — for fresh obligations, recomputed vs stored aggregates
//!      should match (proves the recompute math against on-chain truth).
//!   2. ALPHA — obligations that are liquidatable at current prices, especially
//!      ones whose STORED values still say healthy (stale → a refresh_obligation
//!      would flag them; catching these ahead of the crank is the only edge).
//!
//! Reserve prices come from each reserve's cached market_price (refresh_reserve),
//! which stays fresh because reserves are cranked constantly — so we sidestep
//! Scope. Freshness of those prices is reported.
//!
//! Usage: HELIUS_RPC=<url> [MARKET=all] [MIN_COLLATERAL_USD=50] [NEAR=25]
//!        cargo run --release --bin liq_kamino_live

use arb_engine::kamino::{self, Obligation, Reserve};
use solana_pubkey::Pubkey;
use std::collections::{HashMap, HashSet};
use std::time::Duration;

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() { return Some(v); }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}

fn b64(data: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.decode(data.get(0)?.as_str()?).ok()
}

fn get_multiple(endpoint: &str, keys: &[Pubkey], slice_len: u64) -> HashMap<Pubkey, Vec<u8>> {
    let mut out = HashMap::new();
    for chunk in keys.chunks(100) {
        let strs: Vec<String> = chunk.iter().map(|k| k.to_string()).collect();
        let Some(v) = rpc(endpoint, serde_json::json!({
            "jsonrpc":"2.0","id":1,"method":"getMultipleAccounts",
            "params":[strs, {"encoding":"base64","dataSlice":{"offset":0,"length":slice_len}}]
        })) else { continue };
        for (i, acc) in v["result"]["value"].as_array().into_iter().flatten().enumerate() {
            if let Some(bytes) = acc.get("data").and_then(b64) { out.insert(chunk[i], bytes); }
        }
        std::thread::sleep(Duration::from_millis(40));
    }
    out
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC").or_else(|_| std::env::var("RPC_HTTP"))
        .expect("HELIUS_RPC in .env");
    let market = std::env::var("MARKET").unwrap_or_else(|_| "all".into());
    let min_collateral: f64 = std::env::var("MIN_COLLATERAL_USD").ok().and_then(|s| s.parse().ok()).unwrap_or(50.0);
    let near_n: usize = std::env::var("NEAR").ok().and_then(|s| s.parse().ok()).unwrap_or(25);

    let cur_slot = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getSlot","params":[{"commitment":"confirmed"}]}))
        .and_then(|v| v["result"].as_u64()).unwrap_or(0);

    // 1) Obligations (dataSlice through has_debt @2287).
    let mut filters = vec![serde_json::json!({"dataSize": kamino::OBLIGATION_SIZE})];
    if market != "all" { filters.push(serde_json::json!({"memcmp":{"offset":32,"bytes":market}})); }
    eprintln!("[live] getProgramAccounts obligations …");
    let resp = rpc(&endpoint, serde_json::json!({
        "jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[kamino::KLEND_PROGRAM, {"encoding":"base64","dataSlice":{"offset":0,"length":2288},"filters":filters}]
    }));
    let entries = resp.as_ref().and_then(|v| v["result"].as_array()).cloned().unwrap_or_default();
    let obs: Vec<Obligation> = entries.iter()
        .filter_map(|e| b64(&e["account"]["data"]))
        .filter_map(|b| Obligation::decode(&b))
        .filter(|o| !o.borrows.is_empty())
        .collect();
    eprintln!("[live] {} obligations with debt, current slot {}", obs.len(), cur_slot);
    if obs.is_empty() { return; }

    // 2) Fetch + decode every referenced reserve (need through borrow_factor @5008).
    let reserve_pks: Vec<Pubkey> = obs.iter()
        .flat_map(|o| o.deposits.iter().map(|d| d.0).chain(o.borrows.iter().map(|b| b.0)))
        .collect::<HashSet<_>>().into_iter().collect();
    eprintln!("[live] fetching {} reserves …", reserve_pks.len());
    let reserve_raw = get_multiple(&endpoint, &reserve_pks, 5016);
    let mut reserves: HashMap<Pubkey, Reserve> = HashMap::new();
    for (pk, raw) in &reserve_raw {
        if let Some(r) = Reserve::decode(raw) { reserves.insert(*pk, r); }
    }
    // Reserve price freshness (these cached prices drive the recompute).
    let mut ages: Vec<u64> = reserves.values().map(|r| cur_slot.saturating_sub(r.price_slot)).collect();
    ages.sort_unstable();
    let med_age = ages.get(ages.len()/2).copied().unwrap_or(0);
    eprintln!("[live] decoded {} reserves; cached-price age median {}sl (~{}s), max {}sl",
        reserves.len(), med_age, med_age * 2 / 5, ages.last().copied().unwrap_or(0));

    // Population diagnostics.
    let n_elev = obs.iter().filter(|o| o.elevation_group != 0).count();
    let n_stale = obs.iter().filter(|o| o.stale).count();
    let n_trust = obs.iter().filter(|o| kamino::recompute(o, &reserves).trustworthy()).count();
    eprintln!("[live] population: {} debt obs | {} elevation-group | {} stale | {} trustworthy(non-elev, fully priced)",
        obs.len(), n_elev, n_stale, n_trust);

    // 3) Validation: recomputed vs stored. Compare on trustworthy obligations
    // whose recompute used fresh-enough reserve prices AND that were refreshed
    // recently (so stored ≈ current). Match ⇒ recompute math is correct.
    let mut val_err: Vec<f64> = Vec::new();
    let mut shown = 0;
    println!("\n──── VALIDATION: recomputed vs stored ────");
    for o in &obs {
        let r = kamino::recompute(o, &reserves);
        if !r.trustworthy() || o.unhealthy_borrow_value < 100.0 { continue; }
        // both the obligation and the reserve prices it uses must be recent.
        if cur_slot.saturating_sub(o.last_update_slot) > 300 { continue; }
        if cur_slot.saturating_sub(r.oldest_price_slot) > 300 { continue; }
        let err = (r.unhealthy_borrow_value - o.unhealthy_borrow_value).abs() / o.unhealthy_borrow_value;
        val_err.push(err);
        if shown < 10 {
            println!("  stored unhealthy=${:.2} debt=${:.2} depos=${:.2}  |  recomp unhealthy=${:.2} debt=${:.2} depos=${:.2}  (err {:.2}%)",
                o.unhealthy_borrow_value, o.bf_adjusted_debt, o.deposited_value,
                r.unhealthy_borrow_value, r.bf_adjusted_debt, r.deposited_value, err*100.0);
            shown += 1;
        }
    }
    val_err.sort_by(|a,b| a.partial_cmp(b).unwrap());
    if val_err.is_empty() {
        println!("  (no obligation with both itself + its reserve prices fresh enough to validate)");
    } else {
        println!("  → {} validated, median error {:.3}%, p90 {:.3}%",
            val_err.len(), val_err[val_err.len()/2]*100.0, val_err[(val_err.len()*9/10).min(val_err.len()-1)]*100.0);
    }

    // ALPHA: liquidatable at current prices, ranked by seizable collateral.
    struct Hit<'a> { o: &'a Obligation, r: kamino::Recomputed, stored_liq: bool }
    let mut hits: Vec<Hit> = Vec::new();
    let mut near: Vec<(f64, &Obligation, kamino::Recomputed)> = Vec::new();
    for o in &obs {
        let r = kamino::recompute(o, &reserves);
        if !r.trustworthy() || r.deposited_value < min_collateral { continue; }
        if r.liquidatable() {
            hits.push(Hit { o, r, stored_liq: o.liquidatable() });
        } else if r.ratio() > 0.90 {
            near.push((r.ratio(), o, r));
        }
    }
    hits.sort_by(|a,b| b.r.deposited_value.partial_cmp(&a.r.deposited_value).unwrap_or(std::cmp::Ordering::Equal));
    let hidden_alpha = hits.iter().filter(|h| !h.stored_liq).count();

    println!("\n════ Kamino LIVE liquidatable (recomputed at current prices) ════");
    println!("obligations w/ debt: {}   collateral ≥ ${:.0}", obs.len(), min_collateral);
    println!("LIQUIDATABLE NOW: {}   [{} already flagged by stored values, {} HIDDEN (stored says healthy = stale alpha)]",
        hits.len(), hits.len()-hidden_alpha, hidden_alpha);
    for h in hits.iter().take(40) {
        let age = cur_slot.saturating_sub(h.o.last_update_slot);
        println!("  {} {}…  collateral=${:.2}  debt=${:.2}  thresh=${:.2}  ratio={:.4}  (obl age {}sl{})",
            if h.stored_liq {"known"} else {"ALPHA"}, &h.o.owner.to_string()[..8],
            h.r.deposited_value, h.r.bf_adjusted_debt, h.r.unhealthy_borrow_value, h.r.ratio(),
            age, if h.o.stale {", stale"} else {""});
    }

    near.sort_by(|a,b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
    println!("\nclosest to liquidation at current prices (ratio→1.0):");
    for (ratio, o, r) in near.iter().take(near_n) {
        println!("  {}…  ratio={:.4}  debt=${:.2}  thresh=${:.2}  collateral=${:.2}",
            &o.owner.to_string()[..8], ratio, r.bf_adjusted_debt, r.unhealthy_borrow_value, r.deposited_value);
    }
    println!();
}

File: src/bin/liqKaminoLive.ts
// Port of src/bin/liq_kamino_live.rs
//
// Kamino LIVE-health finder — recomputes each obligation's health from CURRENT
// reserve prices (replicating refresh_obligation), instead of trusting the
// obligation's stored (possibly stale) values.
//
// Two outputs:
//   1. VALIDATION — for fresh obligations, recomputed vs stored aggregates
//      should match (proves the recompute math against on-chain truth).
//   2. ALPHA — obligations that are liquidatable at current prices, especially
//      ones whose STORED values still say healthy (stale -> a refresh_obligation
//      would flag them; catching these ahead of the crank is the only edge).
//
// Reserve prices come from each reserve's cached market_price (refresh_reserve),
// which stays fresh because reserves are cranked constantly — so we sidestep
// Scope. Freshness of those prices is reported.
//
// Usage: HELIUS_RPC=<url> [MARKET=all] [MIN_COLLATERAL_USD=50] [NEAR=25]
//        tsx src/bin/liqKaminoLive.ts

import 'dotenv/config';
import { PublicKey } from '@solana/web3.js';
import {
  decodeObligation,
  decodeReserve,
  KLEND_PROGRAM,
  obligationLiquidatable,
  recompute,
  recomputedLiquidatable,
  recomputedRatio,
  recomputedTrustworthy,
  type Obligation,
  type Recomputed,
  type ReserveMap,
} from '../lib/kamino.js';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await res.json();
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function b64(data: any): Buffer | undefined {
  const s = data?.[0];
  if (typeof s !== 'string') return undefined;
  try {
    return Buffer.from(s, 'base64');
  } catch {
    return undefined;
  }
}

async function getMultiple(endpoint: string, keys: PublicKey[], sliceLen: number): Promise<Map<string, Buffer>> {
  const out = new Map<string, Buffer>();
  for (let i = 0; i < keys.length; i += 100) {
    const chunk = keys.slice(i, i + 100);
    const strs = chunk.map((k) => k.toBase58());
    const v = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getMultipleAccounts',
      params: [strs, { encoding: 'base64', dataSlice: { offset: 0, length: sliceLen } }],
    });
    const arr: unknown[] = Array.isArray(v?.result?.value) ? v.result.value : [];
    arr.forEach((acc, j) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const bytes = (acc as any)?.data !== undefined ? b64((acc as any).data) : undefined;
      if (bytes !== undefined) out.set(chunk[j].toBase58(), bytes);
    });
    await sleep(40);
  }
  return out;
}

interface Hit {
  o: Obligation;
  r: Recomputed;
  storedLiq: boolean;
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (!endpoint) throw new Error('HELIUS_RPC in .env');
  const market = process.env.MARKET ?? 'all';
  const minCollateral = Number.parseFloat(process.env.MIN_COLLATERAL_USD ?? '') || 50.0;
  const nearN = Number.parseInt(process.env.NEAR ?? '', 10) || 25;

  const slotResp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getSlot',
    params: [{ commitment: 'confirmed' }],
  });
  const curSlot: bigint = typeof slotResp?.result === 'number' ? BigInt(slotResp.result) : 0n;

  // 1) Obligations (dataSlice through has_debt @2287).
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const filters: any[] = [{ dataSize: 3344 }];
  if (market !== 'all') filters.push({ memcmp: { offset: 32, bytes: market } });
  console.error('[live] getProgramAccounts obligations …');
  const resp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [KLEND_PROGRAM, { encoding: 'base64', dataSlice: { offset: 0, length: 2288 }, filters }],
  });
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const entries: any[] = Array.isArray(resp?.result) ? resp.result : [];
  const obs: Obligation[] = [];
  for (const e of entries) {
    const bytes = e?.account?.data !== undefined ? b64(e.account.data) : undefined;
    if (bytes === undefined) continue;
    const o = decodeObligation(bytes);
    if (o !== null && o.borrows.length > 0) obs.push(o);
  }
  console.error(`[live] ${obs.length} obligations with debt, current slot ${curSlot}`);
  if (obs.length === 0) return;

  // 2) Fetch + decode every referenced reserve (need through borrow_factor @5008).
  const reservePkSet = new Set<string>();
  const reservePks: PublicKey[] = [];
  for (const o of obs) {
    for (const [r] of o.deposits) {
      if (!reservePkSet.has(r.toBase58())) {
        reservePkSet.add(r.toBase58());
        reservePks.push(r);
      }
    }
    for (const [r] of o.borrows) {
      if (!reservePkSet.has(r.toBase58())) {
        reservePkSet.add(r.toBase58());
        reservePks.push(r);
      }
    }
  }
  console.error(`[live] fetching ${reservePks.length} reserves …`);
  const reserveRaw = await getMultiple(endpoint, reservePks, 5016);
  const reserves: ReserveMap = new Map();
  for (const [pk, raw] of reserveRaw) {
    const r = decodeReserve(raw);
    if (r !== null) reserves.set(pk, r);
  }
  // Reserve price freshness (these cached prices drive the recompute).
  const ages: bigint[] = Array.from(reserves.values()).map((r) => (curSlot > r.priceSlot ? curSlot - r.priceSlot : 0n));
  ages.sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
  const medAge = ages.length > 0 ? ages[Math.floor(ages.length / 2)] : 0n;
  console.error(
    `[live] decoded ${reserves.size} reserves; cached-price age median ${medAge}sl (~${(medAge * 2n) / 5n}s), max ${ages.length > 0 ? ages[ages.length - 1] : 0n}sl`,
  );

  // Population diagnostics.
  const nElev = obs.filter((o) => o.elevationGroup !== 0).length;
  const nStale = obs.filter((o) => o.stale).length;
  const nTrust = obs.filter((o) => recomputedTrustworthy(recompute(o, reserves))).length;
  console.error(
    `[live] population: ${obs.length} debt obs | ${nElev} elevation-group | ${nStale} stale | ${nTrust} trustworthy(non-elev, fully priced)`,
  );

  // 3) Validation: recomputed vs stored. Compare on trustworthy obligations
  // whose recompute used fresh-enough reserve prices AND that were refreshed
  // recently (so stored ~= current). Match => recompute math is correct.
  const valErr: number[] = [];
  let shown = 0;
  console.log('\n──── VALIDATION: recomputed vs stored ────');
  for (const o of obs) {
    const r = recompute(o, reserves);
    if (!recomputedTrustworthy(r) || o.unhealthyBorrowValue < 100.0) continue;
    // both the obligation and the reserve prices it uses must be recent.
    if (curSlot - o.lastUpdateSlot > 300n) continue;
    if (curSlot - r.oldestPriceSlot > 300n) continue;
    const err = Math.abs(r.unhealthyBorrowValue - o.unhealthyBorrowValue) / o.unhealthyBorrowValue;
    valErr.push(err);
    if (shown < 10) {
      console.log(
        `  stored unhealthy=$${o.unhealthyBorrowValue.toFixed(2)} debt=$${o.bfAdjustedDebt.toFixed(2)} depos=$${o.depositedValue.toFixed(2)}  |  recomp unhealthy=$${r.unhealthyBorrowValue.toFixed(2)} debt=$${r.bfAdjustedDebt.toFixed(2)} depos=$${r.depositedValue.toFixed(2)}  (err ${(err * 100.0).toFixed(2)}%)`,
      );
      shown += 1;
    }
  }
  valErr.sort((a, b) => a - b);
  if (valErr.length === 0) {
    console.log('  (no obligation with both itself + its reserve prices fresh enough to validate)');
  } else {
    const p90idx = Math.min(Math.floor((valErr.length * 9) / 10), valErr.length - 1);
    console.log(
      `  → ${valErr.length} validated, median error ${(valErr[Math.floor(valErr.length / 2)] * 100.0).toFixed(3)}%, p90 ${(valErr[p90idx] * 100.0).toFixed(3)}%`,
    );
  }

  // ALPHA: liquidatable at current prices, ranked by seizable collateral.
  const hits: Hit[] = [];
  const near: Array<[number, Obligation, Recomputed]> = [];
  for (const o of obs) {
    const r = recompute(o, reserves);
    if (!recomputedTrustworthy(r) || r.depositedValue < minCollateral) continue;
    if (recomputedLiquidatable(r)) {
      hits.push({ o, r, storedLiq: obligationLiquidatable(o) });
    } else if (recomputedRatio(r) > 0.9) {
      near.push([recomputedRatio(r), o, r]);
    }
  }
  hits.sort((a, b) => b.r.depositedValue - a.r.depositedValue);
  const hiddenAlpha = hits.filter((h) => !h.storedLiq).length;

  console.log('\n════ Kamino LIVE liquidatable (recomputed at current prices) ════');
  console.log(`obligations w/ debt: ${obs.length}   collateral ≥ $${minCollateral.toFixed(0)}`);
  console.log(
    `LIQUIDATABLE NOW: ${hits.length}   [${hits.length - hiddenAlpha} already flagged by stored values, ${hiddenAlpha} HIDDEN (stored says healthy = stale alpha)]`,
  );
  for (const h of hits.slice(0, 40)) {
    const age = curSlot - h.o.lastUpdateSlot;
    console.log(
      `  ${h.storedLiq ? 'known' : 'ALPHA'} ${h.o.owner.toBase58().slice(0, 8)}…  collateral=$${h.r.depositedValue.toFixed(2)}  debt=$${h.r.bfAdjustedDebt.toFixed(2)}  thresh=$${h.r.unhealthyBorrowValue.toFixed(2)}  ratio=${recomputedRatio(h.r).toFixed(4)}  (obl age ${age}sl${h.o.stale ? ', stale' : ''})`,
    );
  }

  near.sort((a, b) => b[0] - a[0]);
  console.log('\nclosest to liquidation at current prices (ratio→1.0):');
  for (const [ratio, o, r] of near.slice(0, nearN)) {
    console.log(
      `  ${o.owner.toBase58().slice(0, 8)}…  ratio=${ratio.toFixed(4)}  debt=$${r.bfAdjustedDebt.toFixed(2)}  thresh=$${r.unhealthyBorrowValue.toFixed(2)}  collateral=$${r.depositedValue.toFixed(2)}`,
    );
  }
  console.log();
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
